libraries/AP_HAL_Linux/RCInput_RPI.cpp

```diff
@@ -476,13 +498,23 @@ void RCInput_RPI::init()
 void RCInput_RPI::_timer_tick()
 {
     uint32_t counter = 0;
+    uint64_t signal_states(0);
 
     if (!_initialized) {
         return;
     }
 
     // Now we are getting address in which DMAC is writing at current moment
     dma_cb_t *ad = (dma_cb_t *)con_blocks->get_virt_addr(dma_reg[RCIN_RPI_DMA_CONBLK_AD | RCIN_RPI_DMA_CHANNEL << 8]);
+
+    if (!ad) {
+        debug("DMA sampling stopped, restarting...\n");
+        init_ctrl_data();
+        init_PCM();
+        init_DMA();
+        return;
```

Comment:
Can you split this commit into things that are exclusive to the PWM and things like the above that are more a safety thing?  This `if` here could be applied very easily, and possibly other parts. Thanks.

Reply:
I divided this commit into three, still, the PWM part changed many lines. I'm trying to make it as easy to review as it can be. If you think something still can be split, please let me know.